[MPS] Prerequisite for MPS C++ extension

qqaatw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102483

in order to add mps kernels to torchvision codebase, we need to expose mps headers and allow objc++ files used in extensions.

cc @kulinseth @albanD

qqaatw committed:
[MPS] Prerequisite for MPS C++ extension

[ghstack-poisoned]

albanD reviewed (commented):
This seems to add extra calls to cudaGetDevice?
Maybe ops being registered from `.mm` files that were not picked up before?
Or a broken trunk?

qqaatw commented:
> This seems to add extra calls to cudaGetDevice?
> 
> Maybe ops being registered from `.mm` files that were not picked up before?
> 
> Or a broken trunk?

It looks like a broken trunk: https://hud.pytorch.org/pr/102483

qqaatw commented:
@kulinseth I proposed a draft PR https://github.com/qqaatw/vision/pull/1 to my torchvision fork (also one in torchvision https://github.com/pytorch/vision/pull/7643), with this setup it works locally on my side (I haven't completed the kernel so the output is incorrect). I think enabling `.mm` is a must, otherwise we cannot compile `.mm` files in torchvision.

I did a quick review on the PR you suggest and am not sure if it is suitable for torchvision codebase, as in torchvision they directly make use of ATen headers/apis (cpu, cuda, hip) that have many convenient utils available, and the PR mentioned seems more suitable for users that just want to write a custom kernel themselves using libtorch rather than deeply integrate their code into PyTorch.

malfet commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -544,7 +544,7 @@ install(FILES "${CMAKE_CURRENT_BINARY_DIR}/cmake-exports/ATenConfig.cmake"
 set(INSTALL_HEADERS ${base_h} ${ATen_CORE_HEADERS})
 if(NOT INTERN_BUILD_MOBILE)
-  list(APPEND INSTALL_HEADERS ${native_h} ${native_cpu_h} ${native_ao_sparse_h} ${native_quantized_h} ${cuda_h} ${native_cuda_h} ${native_hip_h} ${cudnn_h} ${hip_h} ${miopen_h})
+  list(APPEND INSTALL_HEADERS ${native_h} ${native_cpu_h} ${native_ao_sparse_h} ${native_quantized_h} ${cuda_h} ${native_cuda_h} ${native_hip_h} ${cudnn_h} ${hip_h} ${mps_h} ${native_mps_h} ${miopen_h})
```
Can you please elaborate why `native_mps_h` are needed? I.e. aren't those an implementation details?

qqaatw replied:
Sure because I include `ATen/native/mps/OperationUtils.h` in the vision kernels. As a result, we just need this header to have everything work (possibly plus some others like `MPSProfiler.h`).

malfet commented on torch/utils/cpp_extension.py:
```diff
@@ -524,8 +524,8 @@ def build_extensions(self) -> None:
-        # Register .cu, .cuh and .hip as valid source extensions.
-        self.compiler.src_extensions += ['.cu', '.cuh', '.hip']
+        # Register .cu, .cuh, .hip, and .mm as valid source extensions.
+        self.compiler.src_extensions += ['.cu', '.cuh', '.hip', '.mm']
```
Why not guard those by `if sys.platform == 'darwin'` or `if torch.backends.mps.is_built`?

qqaatw replied:
From my understanding the purpose of this list is to allow the extensions to be compiled only. Whether they will actually be compiled is contingent upon whether they're specified as the source files in a Python extension or not. But let's see if the CI complains.

qqaatw replied:
Added `if torch.backends.mps.is_built` guard as a precaution against something possibly failing in the internal builds at Meta.

malfet reviewed (commented):
Please add test (i.e. compile a hellow-world .mm extension or by turning example from #100661 into a real extension)

qqaatw commented:
@malfet is there an extension build test for other backends somewhere that I can take reference? Or where would you suggest to place the test?

malfet commented:
@qqaatw sure, see `test_cpp_extensions_jit.py` and `test_load_standalone` in `test_utils.py` (JIT here is a bit of a misnomer and stands for just-in-time compilation of the extension) and `test/cpp_extensions` folder

qqaatw committed:
Update on "[MPS] Prerequisite for MPS C++ extension"


in order to add mps kernels to torchvision codebase, we need to expose mps headers and allow objc++ files used in extensions.

cc kulinseth albanD 

[ghstack-poisoned]

qqaatw committed:
Update on "[MPS] Prerequisite for MPS C++ extension"


in order to add mps kernels to torchvision codebase, we need to expose mps headers and allow objc++ files used in extensions.

cc kulinseth albanD 

[ghstack-poisoned]

kulinseth commented:
> @kulinseth I proposed a draft PR [qqaatw/vision#1](https://github.com/qqaatw/vision/pull/1) to my torchvision fork (also one in torchvision [pytorch/vision#7643](https://github.com/pytorch/vision/pull/7643)), with this setup it works locally on my side (I haven't completed the kernel so the output is incorrect). I think enabling `.mm` is a must, otherwise we cannot compile `.mm` files in torchvision.
> 
> I did a quick review on the PR you suggest and am not sure if it is suitable for torchvision codebase, as in torchvision they directly make use of ATen headers/apis (cpu, cuda, hip) that have many convenient utils available, and the PR mentioned seems more suitable for users that just want to write a custom kernel themselves using libtorch rather than deeply integrate their code into PyTorch.

I see, I will leave it upto @malfet to comment further from the infrastructure standpoint. But overall it makes sense to me.

qqaatw commented:
@malfet It seems that the mps headers were not installed in the build job, causing the headers not found in the test. (I downloaded the build artifacts and the mps headers do not exist indeed). Do you know what might cause this? it works fine locally.

qqaatw committed:
Update on "[MPS] Prerequisite for MPS C++ extension"


in order to add mps kernels to torchvision codebase, we need to expose mps headers and allow objc++ files used in extensions.

cc kulinseth albanD 

[ghstack-poisoned]

qqaatw committed:
Update on "[MPS] Prerequisite for MPS C++ extension"


in order to add mps kernels to torchvision codebase, we need to expose mps headers and allow objc++ files used in extensions.

cc kulinseth albanD 

[ghstack-poisoned]

qqaatw committed:
Update on "[MPS] Prerequisite for MPS C++ extension"


in order to add mps kernels to torchvision codebase, we need to expose mps headers and allow objc++ files used in extensions.

cc kulinseth albanD 

[ghstack-poisoned]

qqaatw commented on aten/src/ATen/mps/MPSProfiler.h:
```diff
@@ -51,15 +50,7 @@ struct BaseInfo {
   // handle used to identify the profile info's instance (usually the pointer)
   const uintptr_t handle;
 
-  virtual const std::string toString(double gpuTime = 0, double schedulingTime = 0) const {
-    // the gpuTime will be non-zero mainly for event-based signposts.
-    // The interval-based signposts will have "duration" as well as accumulated
-    // total GPU time, up to the point of execution.
-    return fmt::format("{}{}",
-                       gpuTime > 0.0 ? fmt::format(", gpu={:.3f} ms", gpuTime) : "",
-                       schedulingTime > 0.0 ? fmt::format(", cpu={:.3f} ms", schedulingTime) : "");
-  }
-
```
Moved functions that use the fmt lib to `MPSProfiler.mm` since the fmt lib is not packed into the torch wheel, and it thus will not be found in the torchvision c++ extension.

qqaatw committed:
Update on "[MPS] Prerequisite for MPS C++ extension"


in order to add mps kernels to torchvision codebase, we need to expose mps headers and allow objc++ files used in extensions.

cc kulinseth albanD 

[ghstack-poisoned]

malfet reviewed (approved):
Looks good to me, and thank you very much for adding thorough tests

malfet commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge -f "This change has not been changed recently"
